domain/patients/sql_repository.go

```diff
@@ -0,0 +1,291 @@
+	if patient.Email != nil {
+		tmp := *patient.Email
+		email = string(tmp)
+	}
+	if patient.Ssn != nil {
+		ssn = *patient.Ssn
+	}
+	if patient.Dob != nil {
+		dob = patient.Dob.Time
+	}
+	*dbPatient = sqlPatient{
+		ID:         string(patient.PatientID),
+		SSN:        sql.NullString{String: ssn, Valid: ssn != ""},
+		CustomerID: customerID,
+		Dob:        sql.NullTime{Time: dob.UTC(), Valid: !dob.IsZero()},
+		Email:      sql.NullString{String: email, Valid: email != ""},
+		FirstName:  patient.FirstName,
+		Surname:    patient.Surname,
+		Gender:     string(patient.Gender),
+		InternalID: patient.InternalID,
+		Zipcode:    patient.Zipcode,
+	}
+	return nil
+}
+
+// sqlContextGetter is an interface provided both by transaction and standard db connection
+type sqlContextGetter interface {
+	GetContext(ctx context.Context, dest interface{}, query string, args ...interface{}) error
+	SelectContext(ctx context.Context, dest interface{}, query string, args ...interface{}) error
+}
+
+type SQLitePatientRepository struct {
+	factory Factory
+	db      *sqlx.DB
+}
+
+const schema = `
+	CREATE TABLE IF NOT EXISTS patient (
+		id TEXT NOT NULL,
+		ssn varchar(20),
+		customer_id varchar(100) NOT NULL,
+		date_of_birth DATETIME DEFAULT NULL,
+		email  varchar(100),
+		first_name varchar(100) NOT NULL DEFAULT '',
+		surname varchar(100) NOT NULL DEFAULT '',
+		gender varchar(10) NOT NULL DEFAULT 'unknown',
+		internal_id varchar(100) NOT NULL,
+		zipcode varchar(10) NOT NULL DEFAULT "",
+		PRIMARY KEY (customer_id, id)
+	);
+`
+
+func NewSQLitePatientRepository(factory Factory, db *sqlx.DB) *SQLitePatientRepository {
+	if db == nil {
+		panic("missing db")
+	}
+
+	db.MustExec(schema)
+
+	return &SQLitePatientRepository{factory: factory, db: db}
+}
+
+func (r SQLitePatientRepository) FindByID(ctx context.Context, customerID, id string) (*domain.Patient, error) {
+	return r.getPatient(ctx, r.db, customerID, id)
+}
+
+func (r SQLitePatientRepository) Update(ctx context.Context, customerID, id string, updateFn func(c domain.Patient) (*domain.Patient, error)) (patient *domain.Patient, err error) {
+	tx, err := r.db.Beginx()
+	if err != nil {
+		return nil, fmt.Errorf("%w, unable to start transaction", err)
+	}
+
+	defer func() {
+		if err == nil {
+			err = tx.Commit()
+		} else {
+			log.Debug(err)
+			tx.Rollback()
+			patient = nil
+		}
+		if err != nil {
+			patient = nil
+		}
+	}()
+
+	patient, err = r.getPatient(ctx, tx, customerID, id)
+	if err != nil {
+		return
+	}
+	updatedPatient, err := updateFn(*patient)
+	if err != nil {
+		return
+	}
+
+	dbPatient := sqlPatient{}
+	err = dbPatient.UnmarshalFromDomainPatient(customerID, *updatedPatient)
+	if err != nil {
+		return
+	}
+
+	const query = `
+	UPDATE patient SET
+		ssn = :ssn,
+		customer_id = :customer_id,
+		date_of_birth = :date_of_birth,
+		email = :email,
+		first_name = :first_name,
+		surname = :surname,
+		gender = :gender, 
+		internal_id = :internal_id,
+		zipcode = :zipcode
+	WHERE id = :id
```

Comment:
create table statement has no unique on ID so this can alter multiple patients across multiple customers. So either:

- make a unique index on patient.ID
- also specify customerID in the update statement (and maybe add `LIMIT 1`)